### PR TITLE
Rustup

### PR DESCRIPTION
**Pull request by flip1995**
r? @ghost

changelog: none

### CODE REVIEW

File: tests/ui-internal/author.rs
Comment (flip1995) [inline tests/ui-internal/author.rs:2]: @blyxyas Can you shed some light on why the `author` lint tests were moved to `ui-internal` and this `warn` was added? (This lint never emits a diagnostic, so `warn` on it doesn't make sense)

File: clippy_lints/src/utils/author.rs
Comment (flip1995) [inline clippy_lints/src/utils/author.rs:1]: The `dump_hir` "lint" definitely needs the same treatment. And I also think the `format_args_collector` needs it. But before adding that to those, I'd like to get input from @blyxyas . The `#[clippy::author]` and `#[clippy::dump]` attributes currently do not work on nightly: [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9caf525b88d0cfe5e428513f61ee04fc)

File: tests/ui/no_lints.rs
Comment (flip1995) [inline tests/ui/no_lints.rs:1]: Also: what is tested with this? I'd like to add a comment explaining this.

**Comment by samueltardieu** [comment]
@flip1995 It would  be great if you could cherrypick the top commit of https://github.com/samueltardieu/rust-clippy/tree/push-zpsvvwnnpylu (commit 83c3a373c945c4fbaba99bcdc58ef2eaa8f6c5da). It fixes match ergonomics in the new commits, just as they have been fixed for the current Clippy sources already a few days ago.

With this patch, and the two PRs waiting, we stay good for Rust 2024.

**Comment by flip1995** [comment]
Done! Thanks for paying attention to this!

File: clippy_lints/src/utils/internal_lints/lint_without_lint_pass.rs
Comment (blyxyas) [inline clippy_lints/src/utils/internal_lints/lint_without_lint_pass.rs:1]: On line 24, `LintPass::get_lints()` should probably be update to `LintPass::lint_vec`. Also, there exists a possibility that the lint may not even be run if it doesn't implement `LintPass::lint_vec` (there are exactly two lintpasses without lints attached to them, and I'm investigating them to see where do they come from. For now they are included, but they may be discarded in the future)

**Review by blyxyas — COMMENTED** [review]
Just a typo from copy-pasting and a nit, everything else is correct in the five commits you listed =^w^=

**Comment by flip1995** [comment]
Thanks for the review. Could you also address my [two](https://github.com/rust-lang/rust-clippy/pull/13639#discussion_r1827003264) [comments](https://github.com/rust-lang/rust-clippy/pull/13639#discussion_r1827003757) above, please?

File: clippy_lints/src/utils/author.rs
Comment (blyxyas) [reply clippy_lints/src/utils/author.rs:1]: I think that this comment is outdated? Making `AUTHOR` into a Lint-less pass was a decision made because we kept [lintless passes as `eval_always` implicitly](https://github.com/rust-lang/rust/pull/125116/commits/71b4d108c7e71c15c0f61d737ebdc0e1cf559d3c#diff-d7e729ebbc081cc3a5881b3365c50fea5ace147c1ef19bf046a797a21780e443R379-R381). This decision was later phased out in pro of making `author` a simple `eval_always` lint.

I'm looking into why `clippy::author` is not working, could it be because Rust has a LintLess author with a nightly Rust (which filters out lintless passes)?

Comment (flip1995) [reply clippy_lints/src/utils/author.rs:1]: Not outdated no. If you copy the code in the playground to a local `test.rs` file and then run `cargo dev lint test.rs` with this branch, the `author` output is emitted (because the added `eval_always` lint), but the HIR dump is not emitted. It seems like the implicit eval_always is not working.

But if you're saying that this is the intended way, I would revert the dummy AUTHOR lint and rather fix that in rustc. 

But that leaves the question: Why is an explicit lint for CTFE required?

File: tests/ui/no_lints.rs
Comment (flip1995) [reply tests/ui/no_lints.rs:1]: Ok, I'll remove it during the sync

File: clippy_lints/src/utils/author.rs
Comment (flip1995) [reply clippy_lints/src/utils/author.rs:1]: Oh: does the same logic need to be added in the `late_lint_crate` function?

https://github.com/blyxyas/rust/blob/71b4d108c7e71c15c0f61d737ebdc0e1cf559d3c/compiler/rustc_lint/src/late.rs#L446-L449

**Review by blyxyas — APPROVED** [review]

File: clippy_utils/src/qualify_min_const_fn.rs
Comment (flip1995) [reply clippy_utils/src/qualify_min_const_fn.rs:382]: Isn't someone implementing this lint right now? :thinking:

Comment (samueltardieu) [reply clippy_utils/src/qualify_min_const_fn.rs:382]: You must be thinking of https://github.com/rust-lang/rust-clippy/pull/11796  which is currently in FCP, but I am not sure it triggers for `.map_or(true, …)`.

Comment (flip1995) [reply clippy_utils/src/qualify_min_const_fn.rs:382]: Right. And Kevin suggested to extend it to `map_or(true` on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/FCP.20.60unnecessary_map_or.60/near/480334139).

File: clippy_lints/src/utils/author.rs
Comment (blyxyas) [reply clippy_lints/src/utils/author.rs:1]: This is intended and the logic needs to be readded in `late_lint_crate` and/or `late_lint_mod`. CTFE  does not need to be a lint, but it is because lint-less passes are being filtered (which I'll correct today).

I'm not sure how to do this, we don't want under any circumstance to get a nightly version that has no CTFE lint/pass, but opening another PR that modifies Clippy would not be ideal AFAIK. Maybe waiting for approval both in both repos and merging them in the same day?

Comment (flip1995) [reply clippy_lints/src/utils/author.rs:1]: Ok, can you open a rustc PR fixing the behavior for lint passes without lints in rustc, confirm that the author and dump_hir "lints" work? Maybe remove the CTFE lint and add an empty lint pass for that too. 

This can all be done in a PR to the rust repo, instead of the Clippy repo.

After that I will redo this sync. Don't worry about that.

Comment (blyxyas) [reply clippy_lints/src/utils/author.rs:1]: https://github.com/rust-lang/rust/pull/132637 was just open. As there's no harm in CTFE still being a Lint, we can wait until the next sync or do it in this sync

**Pull request merged**
